### PR TITLE
HRIN-606: pretix exporters

### PR DESCRIPTION
**Pull request by rimi-itk**
https://jira.itkdev.dk/browse/HRIN-606

Adds route for listing and running predefined pretix event data exporters (cf. https://docs.pretix.eu/en/latest/api/resources/exporters.html)

### CODE REVIEW

File: src/Controller/PretixEventExportersController.php [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+          file_put_contents($filePath, (string) $response->getBody());
+          file_put_contents($filePath . '.headers', json_encode($response->getHeaders()));
+
+          return new RedirectResponse(file_create_url($url));
+        }
+
+        return $response;
+
+      case Response::HTTP_CONFLICT:
+        sleep(2);
```
Comment (turegjorup) [inline src/Controller/PretixEventExportersController.php:170]: Why sleep here?

File: src/Controller/PretixEventExportersController.php [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    // @see https://docs.pretix.eu/en/latest/api/resources/exporters.html#downloading-the-result
+    switch ($response->getStatusCode()) {
+      case Response::HTTP_OK:
+        $this->session->remove($key);
+
+        $header = $response->getHeaderLine('content-disposition');
+        if (preg_match('/filename="(?<filename>[^"]+)"/', $header, $matches)) {
+          $filename = $matches['filename'];
+
+          $url = 'private://itk_pretix/exporters/' . $filename;
```
Comment (turegjorup) [inline src/Controller/PretixEventExportersController.php]: `private://itk_pretix/exporters/` is used in the `Manager`as well. Consider making it a public constant on one of the classes

File: src/Controller/PretixEventExportersController.php [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+          $directory = dirname($url);
+          /** @var \Drupal\Core\File\FileSystem $fileSystem */
+          $fileSystem = \Drupal::service('file_system');
+          $fileSystem->prepareDirectory($directory, FileSystem::CREATE_DIRECTORY);
+          $filePath = $fileSystem->realpath($url);
+          file_put_contents($filePath, (string) $response->getBody());
+          file_put_contents($filePath . '.headers', json_encode($response->getHeaders()));
```
Comment (turegjorup) [inline src/Controller/PretixEventExportersController.php]: You have the Drupal filesystem service in `$fileSystem` - why not use that?

File: src/Exporter/Manager.php [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  /**
+   * Implementation of itk_pretix_file_download.
+   *
+   * @param string $uri
+   *   The file uri.
+   */
+  public function fileDownload(string $uri) {
+    if (0 === strpos($uri, 'private://itk_pretix/exporters/')) {
+      $user = \Drupal::currentUser();
+      if ($user->isAuthenticated()) {
```
Comment (turegjorup) [inline src/Exporter/Manager.php]: Is this the only check? Do we not require a special role to be authorized to download?

File: src/Access/AccessCheck.php [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+  /**
+   * {@inheritdoc}
+   */
+  public function access(AccountInterface $account, RouteMatchInterface $routeMatch) {
+    $node = $routeMatch->getParameter('node');
+    if ($node instanceof NodeInterface && $account->isAuthenticated()) {
+      $user = User::load($account->id());
+      if ($node->access('update', $user)) {
+        return AccessResult::allowed();
```
Comment (turegjorup) [inline src/Access/AccessCheck.php]: Not sure what the access rules are in the business domain, but: 
- why not use `my_module.permissions.yml` to set up proper permissions?
- this doesn't seem in sync with the access control for downloading files in `Manager->fileDownload()`

File: itk_pretix.services.yml [hunk 6]
```diff
@@ -15,3 +15,31 @@ services:
     class: Drupal\itk_pretix\Pretix\OrderHelper
     arguments:
       - '@database'
+
+  itk_pretix.exporter_manager:
+    class: Drupal\itk_pretix\Exporter\Manager
+    tags:
+      - { name: service_collector, tag: itk_pretix_exporter, call: addEventExporter }
```
Comment (turegjorup) [inline itk_pretix.services.yml:27]: All services defined here could be defined as lazy given they only seem to be used bu logged in editors?

**Review by turegjorup — COMMENTED** [review]
* You are using PHP 7.2 features. Consider adding `"php": "^7.2"` to `composer.json.
* Standalone module, but seems to have some hard expectations around access rules and node access?
* `README.md` should be updated with a description of the export feature

File: src/Controller/PretixEventExportersController.php (hunk 1 above)
Comment (rimi-itk) [reply src/Controller/PretixEventExportersController.php:170]: We have to wait for the exporter to finish. I've added a clarifying comment.

File: src/Controller/PretixEventExportersController.php (hunk 2 above)
Comment (rimi-itk) [reply src/Controller/PretixEventExportersController.php]: I've cleaned up handling of saving and loading exporter result files.

File: src/Controller/PretixEventExportersController.php (hunk 3 above)
Comment (rimi-itk) [reply src/Controller/PretixEventExportersController.php]: Fixed.

File: src/Exporter/Manager.php (hunk 4 above)
Comment (rimi-itk) [reply src/Exporter/Manager.php]: I've moved the access check to `AccessCheck::canRunExport` to keep the check in a single place.

File: src/Access/AccessCheck.php (hunk 5 above)
Comment (rimi-itk) [reply src/Access/AccessCheck.php]: For now, any user that can update a node can run exporters and download the results. We may have to refine and improve this in the future.

See my comment https://github.com/itk-dev/itk_pretix_d8/pull/28#discussion_r533205105

**Review by turegjorup — APPROVED** [review]

**Review by turegjorup — APPROVED** [review]

**Pull request merged**